client.go

```diff
@@ -0,0 +1,155 @@
+/*
+
+Package Client is a Go client for the NHS FHIR API.
+
+For more information about the API visit this link here:
+https://digital.nhs.uk/developer/api-catalogue/personal-demographics-service-fhir#api-description__overview
+
+Usage:
+
+You use this library by creating a new client and calling methods on the client.
+
+Example:
+
+	package main
+
+	import (
+		"github.com/welldigital/nhs-fhir"
+		"fmt"
+	)
+
+	func main() {
+		c, err := client.NewClient()
+		if err != nil {
+			panic(err)
+		}
+
+		patient, err := c.Patient.Get("9000000009")
+	}
+
+*/
+
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+type Client struct {
+	BaseURL   *url.URL
+	UserAgent string
+
+	httpClient *http.Client
+
+	common  service // Reuse a single struct instead of allocating one for each service on the heap.
+	Patient *PatientService
+}
+
+//go:generate moq -out client_moq.go . IClient
+type IClient interface {
+	NewRequest(method, path string, body interface{}) (*http.Request, error)
+	Do(ctx context.Context, req *http.Request, v interface{}) (*Response, error)
+}
+
+var errNonNilContext = errors.New("context must be non-nil")
+
+const (
+	defaultBaseURL = "https://sandbox.api.service.nhs.uk/personal-demographics/FHIR/R4/"
+)
+
+// NewClient returns a new FHIR client. If a nil httpClient is provided then a new http.client will be used.
+// To use API methods requiring auth then provide a http.Client which will perform the authentication for you e.g. oauth2
+func NewClient(httpClient *http.Client) *Client {
+	if httpClient == nil {
+		netTransport := &http.Transport{
+			Dial: (&net.Dialer{
+				Timeout: 5 * time.Second,
+			}).Dial,
+			TLSHandshakeTimeout: 5 * time.Second,
+		}
+		httpClient = &http.Client{
+			Timeout:   time.Second * 10,
+			Transport: netTransport,
+		}
+	}
+	baseURL, _ := url.Parse(defaultBaseURL)
+
+	// adds trailing slash
+	if !strings.HasSuffix(baseURL.Path, "/") {
+		baseURL.Path += "/"
+	}
+
+	c := &Client{
+		BaseURL:    baseURL,
+		httpClient: httpClient,
+	}
+
+	c.common.client = c
+
+	// repeat this pattern per service...
+	c.Patient = (*PatientService)(&c.common)
+
+	return c
+}
+
+// NewRequest creates an API request. A relative URL can be provided in path,
+// in which case it is resolved relative to the BaseURL of the Client.
+// Relative URLs should always be specified without a preceding slash. If
+// specified, the value pointed to by body is JSON encoded and included as the
+// request body.
+func (c *Client) NewRequest(method, path string, body interface{}) (*http.Request, error) {
```

Comment:
I don't think we should make this public

Reply:
I think I agree with you here. Like you said previously we can make it private now and if in the future we do find a use case for it then potentially expose it.